cbor2/encoder.py

```diff
@@ -253,20 +284,26 @@ class CBOREncoder(object):
     :param default: a callable that is called by the encoder with three arguments
         (encoder, value, file object) when no suitable encoder has been found, and should use the
         methods on the encoder to encode any objects it wants to add to the data stream
+    :param canonical: Forces mapping types to be output in a stable order to guarantee that the
+        output will always produce the same hash given the same input. This adds an additional
+        restriction on the datatypes that can be used as keys. In this case only strings, bytes
```

Comment:
Where does this restriction come from? Certainly not the RFC.

Reply:
Using floats as mapping keys is ambiguous, the spec handwaves it away in the section about Canonical encoding. See: https://www.rfc-editor.org/errata_search.php?rfc=7049

Particularly when length of encoding is more significant for key ordering than magnitude.

Is it a good idea to extend map encoding to every hashable type?

Reply:
Theoretically, when floats are encoded canonically, they should work as keys. Can you accurately do key lookups for floats? Probably not, but it will work for iteration. The RFC does not forbid it so I don't see why we should.

Reply:
I will think about this more. Look what I can do in python 3.5:

```python
>>> d = { 7.0: "7", 33.3: "33.3" }
>>> d[7]
'7'
>>> d[7.0]
'7'
>>> d[33]
Traceback (most recent call last):
  File "<pyshell#168>", line 1, in <module>
    d[33]
KeyError: 33
>>> d[33.3]
'33.3'
>>> d[33.2999999999999999999999999999999999999999999999999999]
'33.3'
>>> d[33.299999999999999999999999999999]
'33.3'
>>> d[33.299999999999]
Traceback (most recent call last):
  File "<pyshell#172>", line 1, in <module>
    d[33.299999999999]
KeyError: 33.299999999999
>>> 
```